### PR TITLE
fix: [CDS-75585]: sync instance stats iterator run time to separate table

### PR DESCRIPTION
**Pull request by RishabhGupta34**
## Describe your changes
Added a new table to sync the instance stats iterator run time, to store when the iterator ran last. This is to fix an issue when the instances went down to zero and came back up after 1 day interval. 
Let's say a deployment was done on 2nd August and 2 instances came up, now if they went down to zero they won't be synced to `ng_instance_stats` table with 0 count. Now if they came back up on 4th August, the iterator will still check for instances between 2nd to 3rd Aug. as there is a limit of 48 entries to be synced(every 30 mins) meaning 1 day.
So the new instances won't be synced, this causes a discrepancy in the custom dashboards.

Now we will sync the time at which the iterator ran, which will sync the last time it ran, and when fetching the records we will take the max of both this time and the time coming from `ng_instance_stats` table, to prevent any corner case.

Tested locally:
1. Instances were down for a service from past 2 months, and weren't being synced to the `ng_instance_stats` table, brought them up and the iterator started syncing them back again.

## Checklist
- [ ] I've documented the changes in the PR description.
- [ ] I've tested this change either in PR or local environment.
- [ ] If hashcheck failed I followed [the checklist](https://harness.atlassian.net/wiki/spaces/DEL/pages/21016838831/PR+Codebasehash+Check+merge+checklist) and updated this PR description with my answers to make sure I'm not introducing any breaking changes.

## Comment Triggers
<details>
  <summary>Build triggers</summary>

- Feature build: `trigger feature-build`
- 
  Specific builds
  - `trigger manager`
  - `trigger dms`
  - `trigger ng_manager`
  - `trigger cvng `
  - `trigger cmdlib`
  - `trigger template_svc`
  - `trigger events_fmwrk_monitor`
  - `trigger event_server`
  - `trigger change_data_capture`
  -  Trigger multiple builds together: For eg: `trigger dms, manager`
- Immutable delegate `trigger publish-delegate`
</details>

## [Latest PR Check Triggers](https://github.com/harness/harness-core/blob/develop/.github/pull_request_template.md)

<details>
  <summary>PR Check triggers</summary>
You can run multiple PR check triggers by comma separating them in a single comment. e.g. `trigger ut0, ut1`

- Compile: `trigger compile`
- CodeformatCheckstyle: `trigger checkstylecodeformat`
  - CodeFormat: `trigger codeformat`
  - Checkstyle: `trigger checkstyle`
- MessageMetadata: `trigger messagecheck`
- File-Permission-Check: `trigger checkpermission`
- Recency: `trigger recency`
- BuildNumberMetadata: `trigger buildnum`
- Trigger CommonChecks: `trigger commonchecks`
- PMD: `trigger pmd`
- Copyright Check: `trigger copyrightcheck`
- Feature Name Check: `trigger featurenamecheck`
- UnitTests-ALL: `trigger uts`
- UnitTests-0: `trigger ut0`
- UnitTests-1: `trigger ut1`
- UnitTests-2: `trigger ut2`
- UnitTests-3: `trigger ut3`
- UnitTests-4: `trigger ut4`
- UnitTests-5: `trigger ut5`
- UnitTests-6: `trigger ut6`
- UnitTests-7: `trigger ut7`
- UnitTests-8: `trigger ut8`
- UnitTests-9: `trigger ut9`
- CodeBaseHash: `trigger codebasehash`
- CodeFormatCheckstyle: `trigger checkstylecodeformat`
- SonarScan: `trigger ss`
- GitLeaks: `trigger gitleaks`
- Trigger all Checks: `trigger smartchecks`
- Go Build: `trigger gobuild`
- Validate_Reviews: `trigger review`
- Module Dependency Check: `trigger mdc`
</details>

## PR check failures and solutions
https://harness.atlassian.net/wiki/spaces/BT/pages/21106884744/PR+Checks+-+Failures+and+Solutions

## [Contributor license agreement](https://github.com/harness/harness-core/blob/develop/CONTRIBUTOR_LICENSE_AGREEMENT.md)

- - -
This change is [Reviewable](https://reviewable.io/reviews/harness/harness-core/51337)

### CODE REVIEW

**Comment by RishabhGupta34** [comment]
trigger uts

**Comment by RishabhGupta34** [comment]
trigger uts

**Comment by RishabhGupta34** [comment]
trigger ut5

**Comment by RishabhGupta34** [comment]
trigger messagecheck

**Comment by RishabhGupta34** [comment]
trigger messagecheck

**Review by abhinavmittal-harness — DISMISSED** [review]
> Reviewed 16 of 16 files at r1, 1 of 1 files at r2, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/harness/harness-core/51337)* status: 17 file(s) to review. 1 of 2 approvals obtained.

File: 126-instance/src/main/java/io/harness/repositories/instancestatsiterator/InstanceStatsIteratorRepositoryImpl.java [hunk 1]
```diff
@@ -0,0 +1,198 @@
...
+   * @param projectId project of service
+   * @param serviceId identifier of service
+   * @return statement to get instance stats by using service ref and scoped query
+   * @throws SQLException exception
+   */
+  private PreparedStatement getScopedStatement(
+      Connection dbConnection, String accountId, String orgId, String projectId, String serviceId) throws SQLException {
+    IdentifierRef serviceIdentifierRef =
+        IdentifierRefHelper.getIdentifierRefWithScope(accountId, orgId, projectId, serviceId);
+    PreparedStatement statement = null;
```
Comment (vikyathharekal) [inline 126-instance/src/main/java/io/harness/repositories/instancestatsiterator/InstanceStatsIteratorRepositoryImpl.java:121]: It's a good practice to close the PreparedStatement once we are done. Or wrap it in try()

File: 126-instance/src/main/java/io/harness/repositories/instancestatsiterator/InstanceStatsIteratorRepositoryImpl.java [hunk 2]
```diff
@@ -0,0 +1,198 @@
...
+                 getScopedStatementForUpdate(dbConnection, accountId, orgId, projectId, serviceId, timestamp)) {
+          statement.execute();
+          successfulOperation = true;
+        } catch (SQLException ex) {
+          if (totalTries == MAX_RETRY_COUNT) {
+            log.error("Error while updating timestamp for instance stats iterator after all retries");
+          }
+          log.warn("Could not update timestamp for instance stats iterator. Retrying again. Retry number: {}",
+              totalTries, ex);
+          totalTries++;
```
Comment (vikyathharekal) [inline 126-instance/src/main/java/io/harness/repositories/instancestatsiterator/InstanceStatsIteratorRepositoryImpl.java:86]: Should we sleep for some time before hitting the DB again? Maybe an exponential backoff?

File: 126-instance/src/main/java/io/harness/repositories/instancestatsiterator/InstanceStatsIteratorRepositoryImpl.java [hunk 3]
```diff
@@ -0,0 +1,198 @@
...
+          log.error("Error while updating timestamp for instance stats iterator", ex);
+          totalTries++;
+        }
+      }
+    }
+  }
+
+  // ------------------------------- PRIVATE METHODS ------------------------------
+  private InstanceStatsIterator parseInstanceStatsIteratorRecord(ResultSet resultSet) throws SQLException {
+    while (resultSet != null && resultSet.next()) {
```
Comment (vikyathharekal) [inline 126-instance/src/main/java/io/harness/repositories/instancestatsiterator/InstanceStatsIteratorRepositoryImpl.java:97]:
```suggestion
    if (resultSet == null || !resultSet.next()) {
       return null;
    } 
    return InstanceStatsIterator.builder()....build();

```
Looks like we aren't looping as such, do we really need while?

File: 126-instance/src/main/java/io/harness/repositories/instancestatsiterator/InstanceStatsIteratorRepositoryImpl.java [hunk 4]
```diff
@@ -0,0 +1,198 @@
...
+   * @param serviceId identifier of service
+   * @param timestamp timestamp
+   * @return statement to get instance stats by using service ref and scoped query
+   */
+  private PreparedStatement getScopedStatementForUpdate(Connection dbConnection, String accountId, String orgId,
+      String projectId, String serviceId, long timestamp) throws SQLException {
+    IdentifierRef serviceIdentifierRef =
+        IdentifierRefHelper.getIdentifierRefWithScope(accountId, orgId, projectId, serviceId);
+    PreparedStatement statement = null;
+    switch (serviceIdentifierRef.getScope()) {
```
Comment (vikyathharekal) [inline 126-instance/src/main/java/io/harness/repositories/instancestatsiterator/InstanceStatsIteratorRepositoryImpl.java:165]: Is it possible to have a single query string and pass null wherever needed? 
So for PROJECT scope - accountId, and orgId would be null.

File: 126-instance/src/main/java/io/harness/repositories/instancestatsiterator/InstanceStatsIteratorRepositoryImpl.java (hunk 3 above)
Comment (RishabhGupta34) [reply 126-instance/src/main/java/io/harness/repositories/instancestatsiterator/InstanceStatsIteratorRepositoryImpl.java]: removed

File: 126-instance/src/main/java/io/harness/repositories/instancestatsiterator/InstanceStatsIteratorRepositoryImpl.java (hunk 2 above)
Comment (RishabhGupta34) [reply 126-instance/src/main/java/io/harness/repositories/instancestatsiterator/InstanceStatsIteratorRepositoryImpl.java]: added

File: 126-instance/src/main/java/io/harness/repositories/instancestatsiterator/InstanceStatsIteratorRepositoryImpl.java (hunk 4 above)
Comment (RishabhGupta34) [reply 126-instance/src/main/java/io/harness/repositories/instancestatsiterator/InstanceStatsIteratorRepositoryImpl.java:177]: done

**Comment by RishabhGupta34** [comment]
trigger uts

**Comment by RishabhGupta34** [comment]
trigger messagecheck

File: 126-instance/src/main/java/io/harness/repositories/instancestatsiterator/InstanceStatsIteratorQuery.java [hunk 5]
```diff
@@ -0,0 +1,39 @@
...
+      + "ORDER BY REPORTEDAT DESC LIMIT 1"),
+
+  FETCH_LATEST_RECORD_ORG_LEVEL(
+      "SELECT * FROM NG_INSTANCE_STATS_ITERATOR WHERE ACCOUNTID=? AND ORGID=? AND SERVICEID=? "
+      + "ORDER BY REPORTEDAT DESC LIMIT 1"),
+
+  FETCH_LATEST_RECORD_ACCOUNT_LEVEL("SELECT * FROM NG_INSTANCE_STATS_ITERATOR WHERE ACCOUNTID=? AND SERVICEID=? "
+      + "ORDER BY REPORTEDAT DESC LIMIT 1"),
+
+  UPDATE_RECORD_PROJECT_LEVEL(
```
Comment (vikyathharekal) [inline 126-instance/src/main/java/io/harness/repositories/instancestatsiterator/InstanceStatsIteratorQuery.java:26]: nit: `UPDATE_RECORD_SERVICE_LEVEL` or `UPDATE_RECORD`

**Review by vikyathharekal — DISMISSED** [review]

File: 126-instance/src/main/java/io/harness/repositories/instancestatsiterator/InstanceStatsIteratorQuery.java (hunk 5 above)
Comment (RishabhGupta34) [reply 126-instance/src/main/java/io/harness/repositories/instancestatsiterator/InstanceStatsIteratorQuery.java]: done

**Comment by RishabhGupta34** [comment]
trigger uts

**Comment by RishabhGupta34** [comment]
trigger uts

**Comment by RishabhGupta34** [comment]
trigger ut7

**Review by vikyathharekal — APPROVED** [review]

**Review by abhinavmittal-harness — APPROVED** [review]
> Reviewed 1 of 1 files at r4, 2 of 5 files at r5, 2 of 3 files at r6, 2 of 2 files at r7, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/harness/harness-core/51337)* status: 17 file(s) to review. 3. unresolved discussions.

**Comment by RishabhGupta34** [comment]
trigger review

**Pull request merged**